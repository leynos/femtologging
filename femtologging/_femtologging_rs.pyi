from typing import Any as _Any, Literal, Union

FemtoLevel: _Any
LevelName = Literal["TRACE", "DEBUG", "INFO", "WARN", "WARNING", "ERROR", "CRITICAL"]
LevelArg = Union[LevelName, FemtoLevel]

FemtoLogger: _Any
FemtoHandler: _Any
FemtoStreamHandler: _Any
FemtoFileHandler: _Any
FemtoRotatingFileHandler: _Any
StreamHandlerBuilder: _Any
FileHandlerBuilder: _Any
RotatingFileHandlerBuilder: _Any
ConfigBuilder: _Any
LoggerConfigBuilder: _Any
FormatterBuilder: _Any
LevelFilterBuilder: _Any
NameFilterBuilder: _Any
FilterBuildError: type[Exception]
HandlerConfigError: type[Exception]
HandlerIOError: type[Exception]

hello: _Any
get_logger: _Any
reset_manager_py: _Any
