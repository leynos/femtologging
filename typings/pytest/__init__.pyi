from typing import (
    Any,
    Callable,
    ContextManager,
    Pattern,
    Type,
    TypeVar,
    overload,
    ParamSpec,
)

P = ParamSpec("P")
R = TypeVar("R", bound=BaseException)

@overload
def fixture(func: Callable[P, R]) -> Callable[P, R]: ...
@overload
def fixture(
    *,
    scope: str | None = ...,
    autouse: bool | None = ...,
    params: list[Any] | None = ...,
    ids: list[str] | Callable[[Any], str] | None = ...,
    name: str | None = ...,
) -> Callable[[Callable[P, R]], Callable[P, R]]: ...
def raises(
    exc: Type[R],
    match: str | Pattern[str] | None = ...,
    *,
    msg: str | None = ...,
) -> ContextManager[R]: ...

class mark:
    @staticmethod
    def parametrize(
        argnames: str | list[str],
        argvalues: list[Any] | tuple[Any, ...] | list[tuple[Any, ...]],
        *,
        ids: list[str] | Callable[[Any], str] | None = ...,
        indirect: bool | list[str] = ...,
        **kwargs: Any,
    ) -> Callable[[Callable[..., Any]], Callable[..., Any]]: ...
