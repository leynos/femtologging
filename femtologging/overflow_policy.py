from __future__ import annotations

from enum import Enum


class OverflowPolicy(Enum):
    """Behaviour when a handler queue is full."""

    DROP = "drop"
    BLOCK = "block"
    TIMEOUT = "timeout"

    def __str__(self) -> str:
        return self.value
