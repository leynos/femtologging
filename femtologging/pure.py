from __future__ import annotations


class FemtoLogger:
    """Simplistic Python implementation used when the Rust module is missing."""

    def __init__(self, name: str) -> None:
        self.name = name

    def log(self, level: str, message: str) -> str:
        """Return the formatted log message."""
        return f"{self.name} [{level}] {message}"


def hello() -> str:
    """Return a friendly greeting from Python."""
    return "hello from Python"


__all__ = ["FemtoLogger", "hello"]
