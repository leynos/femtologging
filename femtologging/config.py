"""Configuration via loggingâ€‘style dictionaries.

This module implements :func:`dictConfig`, a restricted variant of
``logging.config.dictConfig``. Only a subset of the standard schema is
recognized: ``filters`` sections, handler ``level`` attributes, and
incremental configuration are unsupported.

String level parameters accept case-insensitive names: "TRACE", "DEBUG",
"INFO", "WARN", "WARNING", "ERROR", and "CRITICAL". "WARN" and "WARNING"
are equivalent.

Example
-------
>>> dictConfig({
...     "version": 1,
...     "handlers": {"h": {"class": "femtologging.StreamHandler"}},
...     "root": {"level": "INFO", "handlers": ["h"]},
... })

The ``dictConfig`` format does not support ``filters`` and will raise ``ValueError`` if a ``filters`` section is provided. To attach filters, use the builder API:

    cb = (
        ConfigBuilder()
        .with_filter("lvl", LevelFilterBuilder().with_max_level("INFO"))
        .with_logger("core", LoggerConfigBuilder().with_filters(["lvl"]))
        .with_root_logger(LoggerConfigBuilder().with_level("INFO"))
    )
    cb.build_and_init()
"""

from __future__ import annotations

import ast
from dataclasses import dataclass
from typing import Any, Callable, Final, Mapping, Sequence, cast
import typing as _typing

if _typing.TYPE_CHECKING:
    from ._femtologging_rs import (  # noqa: F401
        LevelFilterBuilder as LevelFilterBuilder,
        NameFilterBuilder as NameFilterBuilder,
    )

from . import _femtologging_rs as rust
from .overflow_policy import OverflowPolicy

rust = cast(Any, rust)
HandlerConfigError: type[Exception] = getattr(rust, "HandlerConfigError", Exception)
HandlerIOError: type[Exception] = getattr(rust, "HandlerIOError", Exception)

StreamHandlerBuilder = rust.StreamHandlerBuilder
FileHandlerBuilder = rust.FileHandlerBuilder
ConfigBuilder = rust.ConfigBuilder
LoggerConfigBuilder = rust.LoggerConfigBuilder
FormatterBuilder = rust.FormatterBuilder
LevelFilterBuilder = rust.LevelFilterBuilder
NameFilterBuilder = rust.NameFilterBuilder


_HANDLER_CLASS_MAP: Final[dict[str, object]] = {
    "logging.StreamHandler": StreamHandlerBuilder,
    "femtologging.StreamHandler": StreamHandlerBuilder,
    "logging.FileHandler": FileHandlerBuilder,
    "femtologging.FileHandler": FileHandlerBuilder,
}


def _evaluate_string_safely(value: str, context: str) -> object:
    """Safely evaluate a string ``value`` using ``ast.literal_eval``."""
    try:
        return ast.literal_eval(value)
    except (ValueError, SyntaxError) as exc:
        raise ValueError(f"invalid {context}: {value}") from exc


def _validate_mapping_type(value: object, name: str) -> Mapping[object, object]:
    """Ensure ``value`` is a mapping and not bytes-like."""
    if isinstance(value, (bytes, bytearray)) or not isinstance(value, Mapping):
        raise ValueError(f"{name} must be a mapping")
    return cast(Mapping[object, object], value)


def _validate_no_bytes(value: object, name: str) -> None:
    """Reject ``bytes`` or ``bytearray`` for ``value``."""
    if isinstance(value, (bytes, bytearray)):
        raise ValueError(f"{name} must not be bytes or bytearray")


def _validate_string_keys(
    mapping: Mapping[object, object], name: str
) -> Mapping[str, object]:
    """Ensure all keys in ``mapping`` are strings."""
    for key in mapping:
        if not isinstance(key, str):
            raise ValueError(f"{name} keys must be strings")
    return cast(Mapping[str, object], mapping)


def _coerce_args(args: object, ctx: str) -> list[object]:
    """Convert ``args`` into a list for handler construction."""
    if isinstance(args, str):
        args = _evaluate_string_safely(args, f"{ctx} args")
    if args is None:
        return []
    _validate_no_bytes(args, f"{ctx} args")
    if not isinstance(args, Sequence):
        raise ValueError(f"{ctx} args must be a sequence")
    return list(cast(Sequence[object], args))


def _coerce_kwargs(kwargs: object, ctx: str) -> dict[str, object]:
    """Convert ``kwargs`` into a dictionary for handler construction."""
    if isinstance(kwargs, str):
        kwargs = _evaluate_string_safely(kwargs, f"{ctx} kwargs")
    if kwargs is None:
        return {}
    mapping = _validate_mapping_type(kwargs, f"{ctx} kwargs")
    mapping = _validate_string_keys(mapping, f"{ctx} kwargs")
    result: dict[str, object] = {}
    for key, value in mapping.items():
        _validate_no_bytes(value, f"{ctx} kwargs values")
        result[key] = value
    return result


def _resolve_handler_class(name: str) -> object:
    """Return the builder class for ``name`` or raise ``ValueError``."""
    cls = _HANDLER_CLASS_MAP.get(name)
    if cls is None:
        raise ValueError(f"unsupported handler class {name!r}")
    return cls


def _validate_handler_keys(hid: str, data: Mapping[str, object]) -> None:
    """Validate that ``data`` contains only supported handler keys."""
    allowed = {"class", "level", "filters", "args", "kwargs", "formatter"}
    unknown = set(data.keys()) - allowed
    if unknown:
        raise ValueError(f"handler {hid!r} has unsupported keys: {sorted(unknown)!r}")


def _validate_handler_class(hid: str, cls_name: object) -> str:
    """Ensure a string handler class name is provided."""
    if not isinstance(cls_name, str):
        raise ValueError(f"handler {hid!r} missing class")
    return cls_name


def _validate_unsupported_features(data: Mapping[str, object]) -> None:
    """Reject handler features not yet implemented."""
    if "level" in data:
        raise ValueError("handler level is not supported")
    if "filters" in data:
        raise ValueError("handler filters are not supported")


def _validate_handler_config(
    hid: str, data: Mapping[str, object]
) -> tuple[str, list[object], dict[str, object], object | None]:
    """Validate handler ``data`` and return construction parameters."""
    _validate_handler_keys(hid, data)
    cls_name = _validate_handler_class(hid, data.get("class"))
    _validate_unsupported_features(data)
    ctx = f"handler {hid!r}"
    args = _coerce_args(data.get("args"), ctx)
    kwargs = _coerce_kwargs(data.get("kwargs"), ctx)
    return cls_name, args, kwargs, data.get("formatter")


def _create_handler_instance(
    hid: str, cls_name: str, args: list[object], kwargs: dict[str, object]
) -> object:
    """Instantiate a handler builder and wrap constructor errors."""
    builder_cls = _resolve_handler_class(cls_name)
    try:
        args_t = tuple(args)
        kwargs_d = dict(kwargs)
        return cast(Any, builder_cls)(*args_t, **kwargs_d)  # pyright: ignore[reportCallIssue]
    except (TypeError, ValueError, HandlerConfigError, HandlerIOError) as exc:
        raise ValueError(f"failed to construct handler {hid!r}: {exc}") from exc


def _build_handler_from_dict(hid: str, data: Mapping[str, object]) -> object:
    """Create a handler builder from ``dictConfig`` handler data."""
    cls_name, args, kwargs, fmt = _validate_handler_config(hid, data)
    builder = cast(Any, _create_handler_instance(hid, cls_name, args, kwargs))
    if fmt is not None:
        if not isinstance(fmt, str):
            raise ValueError("formatter must be a string")
        builder = builder.with_formatter(fmt)
    return builder


def _validate_logger_handlers(handlers_obj: object) -> list[str]:
    """Validate logger ``handlers`` list and return it."""
    if not isinstance(handlers_obj, (list, tuple)):
        raise ValueError("logger handlers must be a list or tuple of strings")
    handlers_seq = cast(Sequence[object], handlers_obj)
    if not all(isinstance(h, str) for h in handlers_seq):
        raise ValueError("logger handlers must be a list or tuple of strings")
    return list(cast(Sequence[str], handlers_seq))


def _validate_logger_config_keys(name: str, data: Mapping[str, object]) -> None:
    """Ensure ``data`` uses only supported logger keys."""
    allowed = {"level", "handlers", "propagate", "filters"}
    unknown = set(data.keys()) - allowed
    if unknown:
        raise ValueError(f"logger {name!r} has unsupported keys: {sorted(unknown)!r}")
    if "filters" in data:
        raise ValueError("filters are not supported")


def _validate_propagate_value(value: object) -> bool:
    """Validate the ``propagate`` value for a logger."""
    if not isinstance(value, bool):
        raise ValueError("logger propagate must be a bool")
    return value


def _build_logger_from_dict(name: str, data: Mapping[str, object]) -> object:
    """Create a ``LoggerConfigBuilder`` from ``dictConfig`` logger data."""
    _validate_logger_config_keys(name, data)
    builder = LoggerConfigBuilder()
    if "level" in data:
        builder = builder.with_level(data["level"])
    if "handlers" in data:
        handlers = _validate_logger_handlers(data["handlers"])
        builder = builder.with_handlers(handlers)
    if "propagate" in data:
        propagate = _validate_propagate_value(data["propagate"])
        builder = builder.with_propagate(propagate)
    return builder


def _validate_dict_config(config: Mapping[str, object]) -> int:
    """Validate top-level configuration and return the version."""
    if "incremental" in config:
        raise ValueError("incremental configuration is not supported")
    version = int(cast(int, config.get("version", 1)))
    if version != 1:
        raise ValueError(f"unsupported configuration version {version}")
    if "filters" in config:
        raise ValueError("filters are not supported")
    return version


def _create_config_builder(version: int, config: Mapping[str, object]) -> object:
    """Initialize a ``ConfigBuilder`` with global options."""
    cb = ConfigBuilder()
    builder = cb.with_version(version)
    if "disable_existing_loggers" in config:
        value = config["disable_existing_loggers"]
        if not isinstance(value, bool):
            raise ValueError("disable_existing_loggers must be a bool")
        builder = builder.with_disable_existing_loggers(value)
    return builder


def _validate_formatter_field(
    fcfg: Mapping[str, object], field: str, field_type: str
) -> str | None:
    """Return the string value for ``field`` or ``None`` if absent."""
    if field not in fcfg:
        return None
    value = fcfg[field]
    if not isinstance(value, str):
        raise ValueError(f"formatter '{field_type}' must be a string")
    return value


def _build_formatter(fcfg: Mapping[str, object]) -> object:
    """Build a :class:`FormatterBuilder` from configuration."""
    allowed = {"format", "datefmt"}
    unknown = set(fcfg.keys()) - allowed
    if unknown:
        raise ValueError(f"formatter has unsupported keys: {sorted(unknown)!r}")
    fb = FormatterBuilder()
    fmt = _validate_formatter_field(fcfg, "format", "format")
    if fmt is not None:
        fb = fb.with_format(fmt)
    datefmt = _validate_formatter_field(fcfg, "datefmt", "datefmt")
    if datefmt is not None:
        fb = fb.with_datefmt(datefmt)
    return fb


def _validate_section_mapping(section: object, name: str) -> Mapping[str, object]:
    """Ensure a configuration ``section`` is a mapping."""
    return cast(Mapping[str, object], _validate_mapping_type(section, name))


@dataclass(frozen=True)
class SectionProcessor:
    """Configuration for :func:`_process_config_section`."""

    section: str
    builder_method: str
    build_func: Callable[[str, Mapping[str, object]], object]
    err_tmpl: str | None = None


def _process_config_section(
    builder: Any, config: Mapping[str, object], processor: SectionProcessor
) -> None:
    """Generic processor for formatter, handler, and logger sections."""
    mapping = cast(
        Mapping[object, object],
        _validate_section_mapping(config.get(processor.section, {}), processor.section),
    )
    method = getattr(builder, processor.builder_method)
    for key, cfg in mapping.items():
        if not isinstance(key, str):
            if processor.err_tmpl is None:
                raise ValueError(f"{processor.section[:-1]} ids must be strings")
            raise ValueError(processor.err_tmpl.format(name=repr(key)))
        method(
            key,
            processor.build_func(
                key,
                _validate_section_mapping(cfg, f"{processor.section[:-1]} config"),
            ),
        )


def _process_formatters(builder: Any, config: Mapping[str, object]) -> None:
    """Attach formatter builders to ``builder``."""
    _process_config_section(
        builder,
        config,
        SectionProcessor(
            "formatters", "with_formatter", lambda fid, m: _build_formatter(m)
        ),
    )


def _process_handlers(builder: Any, config: Mapping[str, object]) -> None:
    """Attach handler builders to ``builder``."""
    _process_config_section(
        builder,
        config,
        SectionProcessor(
            "handlers", "with_handler", lambda hid, m: _build_handler_from_dict(hid, m)
        ),
    )


def _process_loggers(builder: Any, config: Mapping[str, object]) -> None:
    """Attach logger configurations to ``builder``."""
    _process_config_section(
        builder,
        config,
        SectionProcessor(
            "loggers",
            "with_logger",
            lambda name, m: _build_logger_from_dict(name, m),
            err_tmpl="loggers section key {name} must be a string",
        ),
    )


def _process_root_logger(builder: Any, config: Mapping[str, object]) -> None:
    """Configure the root logger."""
    if "root" not in config:
        raise ValueError("root logger configuration is required")
    root = config["root"]
    if not isinstance(root, Mapping):
        raise ValueError("root logger configuration must be a mapping")
    builder.with_root_logger(
        _build_logger_from_dict("root", cast(Mapping[str, object], root))
    )


def dictConfig(config: Mapping[str, object]) -> None:
    """Configure logging using a ``dictConfig``â€‘style dictionary.

    Parameters
    ----------
    config : Mapping[str, object]
        A dictionary compatible with :mod:`logging.config`. Supported keys are
        ``version``, ``disable_existing_loggers``, ``formatters``, ``handlers``,
        ``loggers``, and ``root``. Unsupported features (e.g., ``filters``,
        handler ``level``) raise ``ValueError``.

    Raises
    ------
    ValueError
        If the configuration uses unsupported features or invalid schemas.

    Examples
    --------
    >>> dictConfig({
    ...     "version": 1,
    ...     "handlers": {"h": {"class": "femtologging.StreamHandler"}},
    ...     "root": {"level": "INFO", "handlers": ["h"]},
    ... })
    """

    version = _validate_dict_config(config)
    builder = cast(Any, _create_config_builder(version, config))
    _process_formatters(builder, config)
    _process_handlers(builder, config)
    _process_loggers(builder, config)
    _process_root_logger(builder, config)
    builder.build_and_init()


__all__ = [
    "ConfigBuilder",
    "LoggerConfigBuilder",
    "FormatterBuilder",
    "StreamHandlerBuilder",
    "FileHandlerBuilder",
    "LevelFilterBuilder",
    "NameFilterBuilder",
    "dictConfig",
    "OverflowPolicy",
]
