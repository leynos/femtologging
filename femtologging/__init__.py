"""femtologging package."""

from __future__ import annotations

# Import the Rust extension packaged under this module's namespace first
# to keep imports at the top for linters.
from . import _femtologging_rs as rust
from .overflow_policy import OverflowPolicy
from .config import dictConfig
import logging
import sys
from dataclasses import dataclass
from typing import (
    Callable,
    Iterable,
    TextIO,
    cast,
    overload,
)


hello = rust.hello
FemtoLogger = rust.FemtoLogger
get_logger = rust.get_logger
reset_manager = rust.reset_manager_py
FemtoHandler = rust.FemtoHandler
FemtoStreamHandler = rust.FemtoStreamHandler
FemtoFileHandler = rust.FemtoFileHandler
FemtoRotatingFileHandler = rust.FemtoRotatingFileHandler
HandlerOptions = rust.HandlerOptions
ROTATION_VALIDATION_MSG = rust.ROTATION_VALIDATION_MSG
StreamHandlerBuilder = rust.StreamHandlerBuilder
FileHandlerBuilder = rust.FileHandlerBuilder
RotatingFileHandlerBuilder = rust.RotatingFileHandlerBuilder
ConfigBuilder = rust.ConfigBuilder
LoggerConfigBuilder = rust.LoggerConfigBuilder
FormatterBuilder = rust.FormatterBuilder
LevelFilterBuilder = rust.LevelFilterBuilder
NameFilterBuilder = rust.NameFilterBuilder
FilterBuildError = rust.FilterBuildError
HandlerConfigError = rust.HandlerConfigError
HandlerIOError = rust.HandlerIOError
_force_rotating_fresh_failure = getattr(
    rust, "force_rotating_fresh_failure_for_test", None
)
_clear_rotating_fresh_failure = getattr(
    rust, "clear_rotating_fresh_failure_for_test", None
)

if callable(_force_rotating_fresh_failure) and callable(_clear_rotating_fresh_failure):
    _force_rotating_fresh_failure_for_test = cast(
        Callable[[int, str | None], None],
        _force_rotating_fresh_failure,
    )
    _clear_rotating_fresh_failure_for_test = cast(
        Callable[[], None],
        _clear_rotating_fresh_failure,
    )
else:
    # Feature disabled: expose no-ops that fail loudly when invoked.

    def _force_rotating_fresh_failure_for_test(
        count: int, reason: str | None = None
    ) -> None:
        raise RuntimeError(
            "rotating fresh-failure hook requires the extension built with the 'python' feature"
        )

    def _clear_rotating_fresh_failure_for_test() -> None:
        return


@dataclass
class BasicConfig:
    """Configuration parameters for basicConfig()."""

    level: str | int | None = None
    filename: str | None = None
    stream: TextIO | None = None
    force: bool = False
    handlers: Iterable[FemtoHandler] | None = None


@overload
def basicConfig(config: BasicConfig, /) -> None: ...


@overload
def basicConfig(**kwargs: object) -> None: ...


def basicConfig(config: BasicConfig | None = None, /, **kwargs: object) -> None:
    """Configure the root logger using the builder API.

        Parameters mirror ``logging.basicConfig`` but currently only a subset is
        supported. ``config`` may be a :class:`BasicConfig` instance; if provided,
        its values take precedence over individual parameters. ``level`` may be a
        string or numeric value understood by the standard :mod:`logging` module.
        ``filename`` configures a :class:`FemtoFileHandler`; otherwise a
        :class:`FemtoStreamHandler` targeting ``stderr`` is installed. ``stream``
        may be ``sys.stdout`` to redirect output. ``force`` removes any existing
        handlers from the root logger before applying the new configuration.
        ``handlers`` allows attaching pre‑constructed handlers directly.

        Parameters
        ----------
        config : BasicConfig, optional
            Configuration dataclass providing parameters for ``basicConfig``.

        Other Parameters
        ----------------
        level : str or int, optional
            Logging level. Accepts case-insensitive "TRACE", "DEBUG", "INFO", "WARN",
    "WARNING", "ERROR", and "CRITICAL". "WARN" and "WARNING" are equivalent.
        filename : str, optional
            File to write logs to.
        stream : TextIO, optional
            ``sys.stdout`` or ``sys.stderr``.
        force : bool, default False
            Remove any existing handlers before configuring.
        handlers : Iterable[FemtoHandler], optional
            Pre‑constructed handlers to attach.

        Examples
        --------
        Using a dataclass::

            cfg = BasicConfig(level="INFO")
            basicConfig(cfg)

        Using individual parameters::

            basicConfig(level="INFO")

        Notes
        -----
        ``format`` and ``datefmt`` are intentionally unsupported until formatter
        customisation is implemented.
    """
    allowed = {"level", "filename", "stream", "force", "handlers"}
    unknown = set(kwargs) - allowed
    if unknown:
        name = next(iter(unknown))
        raise TypeError(f"basicConfig() got an unexpected keyword argument {name!r}")

    level: str | int | None
    filename: str | None
    stream: TextIO | None
    force: bool
    handlers: Iterable[FemtoHandler] | None
    if config is not None:
        level = (
            config.level
            if config.level is not None
            else cast(str | int | None, kwargs.get("level"))
        )
        filename = (
            config.filename
            if config.filename is not None
            else cast(str | None, kwargs.get("filename"))
        )
        stream = (
            config.stream
            if config.stream is not None
            else cast(TextIO | None, kwargs.get("stream"))
        )
        force = bool(config.force)
        handlers = (
            config.handlers
            if config.handlers is not None
            else cast(Iterable[FemtoHandler] | None, kwargs.get("handlers"))
        )
    else:
        level = cast(str | int | None, kwargs.get("level"))
        filename = cast(str | None, kwargs.get("filename"))
        stream = cast(TextIO | None, kwargs.get("stream"))
        force = bool(kwargs.get("force", False))
        handlers = cast(Iterable[FemtoHandler] | None, kwargs.get("handlers"))

    _validate_basic_config_params(filename, stream, handlers)

    root = get_logger("root")
    if force:
        root.clear_handlers()

    _configure_handlers(root, handlers, filename, stream)

    _set_logger_level(root, level)


def _has_conflicting_handler_params(
    handlers: Iterable[FemtoHandler] | None,
    filename: str | None,
    stream: TextIO | None,
) -> bool:
    """Check if handlers conflict with filename or stream parameters."""
    return handlers is not None and (filename is not None or stream is not None)


def _validate_basic_config_params(
    filename: str | None,
    stream: TextIO | None,
    handlers: Iterable[FemtoHandler] | None,
) -> None:
    """Validate ``basicConfig`` parameters."""
    if filename and stream:
        raise ValueError("Cannot specify both `filename` and `stream`")

    if _has_conflicting_handler_params(handlers, filename, stream):
        msg = "Cannot specify `handlers` with `filename` or `stream`"
        raise ValueError(msg)

    if stream not in (None, sys.stdout, sys.stderr):
        raise ValueError(
            f"stream must be sys.stdout or sys.stderr, got {type(stream)!r}: {stream!r}"
        )


def _configure_handlers(
    root: FemtoLogger,
    handlers: Iterable[FemtoHandler] | None,
    filename: str | None,
    stream: TextIO | None,
) -> None:
    """Attach or build handlers for the root logger."""
    if handlers is not None:
        for h in handlers:
            root.add_handler(h)
    else:
        _build_and_configure_handler(filename, stream)


def _build_and_configure_handler(
    filename: str | None,
    stream: TextIO | None,
) -> None:
    """Build a handler via the builder API and install it."""
    builder = ConfigBuilder()

    handler_id = "basic_config_handler"
    handler = _create_handler_builder(filename, stream)
    builder.with_handler(handler_id, handler)

    logger_cfg = LoggerConfigBuilder().with_handlers([handler_id])
    builder.with_root_logger(logger_cfg)

    builder.build_and_init()


def _create_handler_builder(
    filename: str | None,
    stream: TextIO | None,
) -> FileHandlerBuilder | StreamHandlerBuilder:
    """Create a handler builder for ``basicConfig``."""
    if filename:
        return FileHandlerBuilder(filename)
    if stream is sys.stdout:
        return StreamHandlerBuilder.stdout()
    return StreamHandlerBuilder.stderr()


def _set_logger_level(root: FemtoLogger, level: str | int | None) -> None:
    """Set the root logger level if provided."""
    if level is not None:
        lvl = logging.getLevelName(level) if isinstance(level, int) else level
        root.set_level(lvl)


__all__ = [
    "FemtoHandler",
    "FemtoLogger",
    "get_logger",
    "reset_manager",
    "FemtoStreamHandler",
    "FemtoFileHandler",
    "FemtoRotatingFileHandler",
    "HandlerOptions",
    "ROTATION_VALIDATION_MSG",
    "StreamHandlerBuilder",
    "FileHandlerBuilder",
    "RotatingFileHandlerBuilder",
    "LevelFilterBuilder",
    "NameFilterBuilder",
    "FilterBuildError",
    "ConfigBuilder",
    "LoggerConfigBuilder",
    "FormatterBuilder",
    "HandlerConfigError",
    "HandlerIOError",
    "OverflowPolicy",
    "BasicConfig",
    "basicConfig",
    "dictConfig",
    "hello",
]
