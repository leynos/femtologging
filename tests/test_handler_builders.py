"""Behaviour-driven tests for FileHandlerBuilder and StreamHandlerBuilder.

Scenarios cover capacity and interval configuration, dictionary snapshots,
and build-time failures.
"""

from __future__ import annotations

from pathlib import Path

import pytest
from pytest_bdd import given, scenarios, then, when, parsers
from syrupy import SnapshotAssertion

from femtologging import (
    FileHandlerBuilder,
    StreamHandlerBuilder,
    HandlerConfigError,
)

scenarios("features/handler_builders.feature")


@given('a FileHandlerBuilder for path "test.log"', target_fixture="file_builder")
def given_file_builder(tmp_path) -> FileHandlerBuilder:
    path = tmp_path / "test.log"
    return FileHandlerBuilder(str(path))


@given("a StreamHandlerBuilder targeting stdout", target_fixture="stream_builder")
def given_stream_stdout() -> StreamHandlerBuilder:
    return StreamHandlerBuilder.stdout()


@given("a StreamHandlerBuilder targeting stderr", target_fixture="stream_builder")
def given_stream_stderr() -> StreamHandlerBuilder:
    return StreamHandlerBuilder.stderr()


@when(parsers.parse("I set file capacity {capacity:d}"))
def when_set_file_capacity(
    file_builder: FileHandlerBuilder, capacity: int
) -> FileHandlerBuilder:
    return file_builder.with_capacity(capacity)


@when(parsers.parse("I set stream capacity {capacity:d}"))
def when_set_stream_capacity(
    stream_builder: StreamHandlerBuilder, capacity: int
) -> StreamHandlerBuilder:
    return stream_builder.with_capacity(capacity)


@when(parsers.parse("I set stream flush timeout {timeout:d}"))
def when_set_stream_flush_timeout(
    stream_builder: StreamHandlerBuilder, timeout: int
) -> StreamHandlerBuilder:
    return stream_builder.with_flush_timeout_ms(timeout)


@when(parsers.parse("I set flush record interval {interval:d}"))
def when_set_flush_record_interval(
    file_builder: FileHandlerBuilder, interval: int
) -> FileHandlerBuilder:
    return file_builder.with_flush_record_interval(interval)


@when("I set overflow policy to timeout with 500ms")
def when_set_overflow_policy_timeout(
    file_builder: FileHandlerBuilder,
) -> FileHandlerBuilder:
    return file_builder.with_overflow_policy("timeout", timeout_ms=500)


@when(parsers.parse('I set file formatter "{formatter_id}"'))
def when_set_file_formatter(
    file_builder: FileHandlerBuilder, formatter_id: str
) -> FileHandlerBuilder:
    return file_builder.with_formatter(formatter_id)


@when(parsers.parse('I set stream formatter "{formatter_id}"'))
def when_set_stream_formatter(
    stream_builder: StreamHandlerBuilder, formatter_id: str
) -> StreamHandlerBuilder:
    return stream_builder.with_formatter(formatter_id)


@then("the file handler builder matches snapshot")
def then_file_builder_snapshot(
    file_builder: FileHandlerBuilder, snapshot: SnapshotAssertion
) -> None:
    data = file_builder.as_dict()
    data["path"] = Path(data["path"]).name
    assert data == snapshot, "file builder dict must match snapshot"
    handler = file_builder.build()
    handler.close()


@then("the file handler builder with timeout overflow matches snapshot")
def then_file_builder_timeout_snapshot(
    file_builder: FileHandlerBuilder, snapshot: SnapshotAssertion
) -> None:
    data = file_builder.as_dict()
    data["path"] = Path(data["path"]).name
    assert data["overflow_policy"] == "timeout", "must record timeout policy"
    assert data["timeout_ms"] == 500, "must record configured timeout"
    assert data == snapshot, "snapshot must include timeout fields"
    handler = file_builder.build()
    handler.close()


@then("building the file handler fails")
def then_file_builder_fails(file_builder: FileHandlerBuilder) -> None:
    with pytest.raises(HandlerConfigError):
        file_builder.build()


@then("the stream handler builder matches snapshot")
def then_stream_builder_snapshot(
    stream_builder: StreamHandlerBuilder, snapshot: SnapshotAssertion
) -> None:
    assert stream_builder.as_dict() == snapshot, (
        "stream builder dict must match snapshot"
    )
    handler = stream_builder.build()
    handler.flush()
    handler.close()


@then("building the stream handler fails")
def then_stream_builder_fails(stream_builder: StreamHandlerBuilder) -> None:
    with pytest.raises(HandlerConfigError):
        stream_builder.build()


@then(parsers.parse("setting stream flush timeout {timeout:d} fails"))
def then_setting_stream_flush_timeout_fails(
    stream_builder: StreamHandlerBuilder, timeout: int
) -> None:
    with pytest.raises(OverflowError):
        stream_builder.with_flush_timeout_ms(timeout)


@pytest.mark.parametrize(
    "ctor", [StreamHandlerBuilder.stdout, StreamHandlerBuilder.stderr]
)
def test_stream_builder_negative_capacity(ctor) -> None:
    builder = ctor()
    with pytest.raises(OverflowError):
        builder.with_capacity(-1)
