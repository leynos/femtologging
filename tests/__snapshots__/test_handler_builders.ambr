# serializer version: 1
# name: test_build_file_handler_builder
  dict({
    'capacity': 10,
    'flush_record_interval': 2,
    'formatter_id': 'default',
    'formatter_kind': 'id',
    'overflow_policy': 'drop',
    'path': 'test.log',
  })
# ---
# name: test_build_rotating_file_handler_builder
  dict({
    'backup_count': 5,
    'capacity': 10,
    'flush_record_interval': 2,
    'formatter_id': 'default',
    'formatter_kind': 'id',
    'max_bytes': 1024,
    'overflow_policy': 'drop',
    'path': 'test.log',
  })
# ---
# name: test_dictconfig_rotating_file_handler_builder
  dict({
    'backup_count': 5,
    'capacity': 10,
    'flush_record_interval': 2,
    'formatter_id': 'default',
    'formatter_kind': 'id',
    'max_bytes': 1024,
    'overflow_policy': 'drop',
    'path': 'test.log',
  })
# ---
# name: test_build_stream_handler_builder
  dict({
    'capacity': 8,
    'formatter_id': 'default',
    'formatter_kind': 'id',
    'target': 'stdout',
  })
# ---
# name: test_build_stream_handler_builder_with_flush_timeout
  dict({
    'flush_timeout_ms': 250,
    'formatter_id': 'default',
    'formatter_kind': 'id',
    'target': 'stdout',
  })
# ---
# name: test_file_handler_builder_with_timeout_overflow_policy
  dict({
    'overflow_policy': 'timeout',
    'path': 'test.log',
    'timeout_ms': 500,
  })
# ---
# name: test_build_socket_handler_builder_for_tcp
  dict({
    'capacity': 8,
    'connect_timeout_ms': 500,
    'host': '127.0.0.1',
    'max_frame_size': 2048,
    'port': 9020,
    'tls': True,
    'tls_domain': 'example.com',
    'tls_insecure': False,
    'transport': 'tcp',
    'write_timeout_ms': 250,
  })
# ---
